src/checkers/inference/model/ConstraintManager.java

```diff
@@ -167,41 +144,92 @@ private AnnotationLocation getCurrentLocation() {
         }
     }
 
+    // All addXXXConstraint methods create a (possibly normalized) constraint for the given slots
+    // and issues errors for unsatisfiable constraints.
+
+    /**
+     * Creates and adds a {@link SubtypeConstraint} between the two slots to the constraint set,
+     * which may be normalized to {@link AlwaysTrueConstraint} or {@link EqualityConstraint}. An
+     * error is issued if the subtype constraint is always unsatisfiable.
+     */
     public void addSubtypeConstraint(Slot subtype, Slot supertype) {
-        if ((subtype instanceof ConstantSlot)
-                && this.realQualHierarchy.getTopAnnotations().contains(((ConstantSlot) subtype).getValue())) {
-            this.addEqualityConstraint(supertype, (ConstantSlot) subtype);
-        } else if ((supertype instanceof ConstantSlot)
-                && this.realQualHierarchy.getBottomAnnotations().contains(
-                        ((ConstantSlot) supertype).getValue())) {
-            this.addEqualityConstraint(subtype, (ConstantSlot) supertype);
-        } else {
-            this.add(this.createSubtypeConstraint(subtype, supertype));
+        Constraint constraint = createSubtypeConstraint(subtype, supertype);
+
+        if (constraint instanceof AlwaysFalseConstraint) {
+            checker.report(Result.failure("subtype.constraint.unsatisfiable", subtype, supertype),
```

Comment:
@topnessman @wmdietl For now, if a `SubtypeConstraint` is normalized to a `AlwaysFalseConstraint`, an error message is raised here but it doesn't crash the inference program, as that's how the `ConstraintManager` currently handles unsatisfiable subtype constraints (https://github.com/opprop/checker-framework-inference/pull/147/files#diff-1194fb72462d89e58d363257999451ceL82).

Should this be a crashing error (as in `ErrorReporter.errorAbort()`) or should add `AlwaysFalseConstraint` into the constraint set for serialization?